src/librustc_mir/const_eval.rs

```diff
@@ -146,6 +146,8 @@ fn eval_body_using_ecx<'mir, 'tcx>(
     let name = ty::tls::with(|tcx| tcx.def_path_str(cid.instance.def_id()));
     let prom = cid.promoted.map_or(String::new(), |p| format!("::promoted[{:?}]", p));
     trace!("eval_body_using_ecx: pushing stack frame for global: {}{}", name, prom);
+    // FIXME (davidhewitt): This assert can trip while const-evaluating a function
+    // where all arguments are zero-sized types
```

Comment:
You can totally remove it. This was just a sanity test so we know if we accidentally evaluate a const fn. Or maybe make the assert do the no args or only zst args check. A build-pass test for such a function call would be good to, so we don't regress it

Reply:
Ok, will see if I can make an assertion for all zst args here. We don't have the type information for the args AFAICT so I need to figure out if we can get it!

Reply:
You can use `tcx.subst_and_normalize_erasing_regions(cid.instance.substs, ParamEnv::reveal_all(), tcx.fn_sig(cid.instance.def_id()).skip_binder()).inputs()` to get all param types.

Reply:
We have the mir body available. You can access the argument types via the local_decls field. There's also an args method, but it may just give you the IDs of the argument locals